File: drone_fronted/vue3-app/src/api/algorithm.js
import api from "./index";
import axios from "axios";

// 获取算法列表
export const getAlgorithms = async (page = 1, limit = 10) => {
    return await api.get(`/algorithms?page=${page}&limit=${limit}`);
};

// 创建算法（处理唯一性错误）
export const createAlgorithm = async (data) => {
    try {
        return await api.post("/algorithms", data);
    } catch (error) {
        if (error.response && error.response.status === 400) {
            console.warn("⚠️ 该算法已存在:", error.response.data.detail);
        } else {
            console.error("创建算法失败:", error);
        }
        throw error;
    }
};

// 删除算法
export const deleteAlgorithm = (id) => api.delete(`/algorithms/${id}`);

export const uploadAlgorithm = async (formData) => {
    const res = await axios.post("/api/upload-algorithm/", formData);
    return res.data;
};

export const saveAlgorithm = async (data) => {
    return await axios.post("/api/save-algorithm/", new URLSearchParams(data));
};

File: drone-backend/app/routers/algorithm.py
from minio_client import upload_file_to_minio
from fastapi import APIRouter, Depends, HTTPException, Query, UploadFile, Form
from sqlalchemy.orm import Session
from database import get_db
from cruds import create_algorithm, get_algorithm, delete_algorithm
from schemas import AlgorithmCreate, AlgorithmResponse
from typing import List
from models import Algorithm
from cruds import save_algorithm_to_db


router = APIRouter(
    prefix="/algorithms",
    tags=["algorithms"]
)

@router.post("/", response_model=AlgorithmResponse)
def create_algorithm_api(algorithm: AlgorithmCreate, db: Session = Depends(get_db)):
    """创建算法（避免 `name + file_path` 重复）"""
    existing_algorithm = db.query(Algorithm).filter(
        Algorithm.name == algorithm.name,
        Algorithm.file_path == algorithm.file_path
    ).first()

    if existing_algorithm:
        raise HTTPException(status_code=400, detail="Algorithm already exists with this name and file_path")

    return create_algorithm(db, algorithm)

@router.get("/", response_model=dict)
def get_algorithms(
    db: Session = Depends(get_db),
    page: int = Query(1, alias="page"),
    limit: int = Query(10, alias="limit"),
):
    """获取分页算法"""
    total = db.query(Algorithm).count()
    items = db.query(Algorithm).offset((page - 1) * limit).limit(limit).all()

    return {
        "total": total,
        "items": [AlgorithmResponse.model_validate(algorithm) for algorithm in items]
    }

@router.get("/{algorithm_id}", response_model=AlgorithmResponse)
def read_algorithm_api(algorithm_id: int, db: Session = Depends(get_db)):
    """根据ID获取算法"""
    algorithm = get_algorithm(db, algorithm_id)
    if not algorithm:
        raise HTTPException(status_code=404, detail="Algorithm not found")
    return algorithm

@router.delete("/{algorithm_id}")
def delete_algorithm_api(algorithm_id: int, db: Session = Depends(get_db)):
    """删除算法"""
    if not delete_algorithm(db, algorithm_id):
        raise HTTPException(status_code=404, detail="Algorithm not found")
    return {"message": "Algorithm deleted successfully"}


@router.post("/upload-algorithm/")
async def upload_algorithm(file: UploadFile):
    """ 上传 Python 代码到 MinIO """
    file_url = upload_file_to_minio(file)
    return {"file_url": file_url}

@router.post("/save-algorithm/")
async def save_algorithm(name: str = Form(...), file_url: str = Form(...)):
    """ 保存算法信息到数据库 """
    algo_id = save_algorithm_to_db(name, file_url)
    return {"message": "Algorithm saved successfully", "id": algo_id}